snarky_curve/snarky_curve.ml

```diff
@@ -447,4 +447,170 @@ module For_native_base_field (Inputs : Native_base_field_inputs) = struct
     |> unshift
 
   let scale_known pc bs = unshift (scale_known pc bs)
+
+  (* b ? t : -t *)
+  let conditional_negation (b : Boolean.var) (x, y) =
+    let y' =
+      exists Field.typ
+        ~compute:
+          As_prover.(
+            fun () ->
+              if read Boolean.typ b then read Field.typ y
+              else Field.Constant.negate (read Field.typ y))
+    in
+    assert_r1cs y Field.((of_int 2 * (b :> Field.t)) - of_int 1) y' ;
+    (x, y')
+
+  let p_plus_q_plus_p ((x1, y1) : t) ((x2, y2) : t) =
+    let open Field in
+    let ( ! ) = As_prover.read typ in
+    let lambda_1 =
+      exists typ ~compute:Constant.(fun () -> (!y2 - !y1) / (!x2 - !x1))
+    in
+    let x3 =
+      exists typ
+        ~compute:Constant.(fun () -> (!lambda_1 * !lambda_1) - !x1 - !x2)
+    in
+    let lambda_2 =
+      exists typ
+        ~compute:
+          Constant.(fun () -> (of_int 2 * !y1 / (!x1 - !x3)) - !lambda_1)
+    in
+    let x4 =
+      exists typ
+        ~compute:Constant.(fun () -> (!lambda_2 * !lambda_2) - !x3 - !x1)
+    in
+    let y4 =
+      exists typ ~compute:Constant.(fun () -> ((!x1 - !x4) * !lambda_2) - !y1)
+    in
+    (* Determines lambda_1 *)
+    assert_r1cs (x2 - x1) lambda_1 (y2 - y1) ;
+    (* Determines x_3 *)
+    assert_square lambda_1 (x1 + x2 + x3) ;
+    (* Determines lambda_2 *)
+    assert_r1cs (x1 - x3) (lambda_1 + lambda_2) (of_int 2 * y1) ;
+    (* Determines x4 *)
+    assert_square lambda_2 (x3 + x1 + x4) ;
+    (* Determines y4 *)
+    assert_r1cs (x1 - x4) lambda_2 (y4 + y1) ;
+    (x4, y4)
+
+  (* Input:
+     t, r (LSB)
+
+     Output:
+    (2*r + 1 + 2^len(r)) t
+  *)
+  let scale (t : Field.t * Field.t) (`Times_two_plus_1_plus_2_to_len r) :
+      Field.t * Field.t =
+    let n = Array.length r in
+    let acc = ref (double t) in
+    let () =
+      for i = 0 to n - 1 do
```

Comment:
This is reversed relative to the https://github.com/zcash/zcash/issues/3924 algorithm; it's probably just that r is MSB-first, but that isn't documented.

(Same in the [current code](https://github.com/o1-labs/snarky/blob/2a85945bb4855484ba076257839bcf77bb65244c/snarky_curve/snarky_curve.ml#L528).)